File: Finance.Domain/Entities/Core/Account.cs
using Finance.Domain.Enums;

namespace Finance.Domain.Entities.Core
{
    public class Account
    {
        public Guid Id { get; set; }

        public string Name { get; set; } = string.Empty;

        public string? AccountNumber { get; set; }

        public string? Description { get; set; }

        public AccountType AccountType { get; set; }

        public decimal OpeningBalance { get; set; }

        public bool IsCore { get; set; }
        public bool IsGloballyShared { get; set; }

        public Guid? OwnerUserId { get; set; }
        public User? OwnerUser { get; set; }

        public AccountReportingMode ReportingMode { get; set; } = AccountReportingMode.Included;

        public Guid? ParentAccountId { get; set; }
        public Account? ParentAccount { get; set; }

        public ICollection<Account> Children { get; set; } = new List<Account>();
        public ICollection<AccountAccess> AccessEntries { get; set; } = new List<AccountAccess>();
        public ICollection<Split> Splits { get; set; } = new List<Split>();
    }
}

File: Finance.WebAPI/finance-spa/models/account.ts
export type AccountType = 1 | 2 | 3 | 4 | 5

export type AccountReportingMode = "Included" | "OperationalOnly" | "Excluded"

export type AccountPermissionSummary = {
  canView: boolean
  canPost: boolean
  canEditTransaction: boolean
  canDeleteTransaction: boolean
  canManageAccess: boolean
  canChangeOwner: boolean
  isOwner: boolean
}

export type AccountAccessEntry = {
  userId: string
  userDisplayName: string
  userEmail: string
  isAdmin: boolean
  canView: boolean
  canPost: boolean
  canEditTransaction: boolean
  canDeleteTransaction: boolean
  canManageAccess: boolean
}

export type AccountPermissionUserOption = {
  id: string
  displayName: string
  email: string
  isAdmin: boolean
}

export type AccountPermissionsDetails = {
  accountId: string
  accountName: string
  ownerUserId: string | null
  ownerDisplayName: string | null
  ownerEmail: string | null
  isGloballyShared: boolean
  reportingMode: AccountReportingMode
  currentUserPermissions: AccountPermissionSummary
  entries: AccountAccessEntry[]
  availableUsers: AccountPermissionUserOption[]
}

export type Account = {
  id: string
  name: string
  accountNumber: string | null
  description: string | null
  accountType: AccountType | string
  parentAccountId: string | null
  openingBalance: number
  isCore: boolean
  isGloballyShared: boolean
  ownerUserId: string | null
  ownerDisplayName: string | null
  ownerEmail: string | null
  reportingMode: AccountReportingMode
  currentUserPermissions: AccountPermissionSummary
}

export type CreateAccountInput = {
  id?: string
  name: string
  accountNumber?: string | null
  description?: string | null
  accountType: AccountType
  parentAccountId: string | null
  openingBalance: number
}

export type AccountNode = Account & {
  children: AccountNode[]
}

export type BatchUpdateAccountsInput = {
  accountIds: string[]
  applyOwner: boolean
  ownerUserId: string | null
  applyGlobalSharing: boolean
  isGloballyShared: boolean
  applyMove: boolean
  parentAccountId: string | null
}

export type BatchUpdateAccountResult = {
  id: string
  name: string
  updatedNodeCount: number
  accountTypeChanged: boolean
}
